### ISSUE TITLE
Copter 3.6 and up : able to import 3D object for avoidance

### ISSUE THREAD

**Issue by roque-canales** [ArduPilot/ardupilot#11343]
Hello,

Idea is to import 3D object for avoidance without sensor.

Just with GPS, for example , I enter the 3D volume of one tree, and copter can't go in.

Is this feature exist ?
If this feature not exist maybe with companion computer it's possible to send 3D objects for avoidance ?

**Comment by rmackay9** [comment]
We don't support this but I think it is very similar to stay-out zones.  The possible addition here is the vertical axis.. stay out zones are generally 2D with infinite height.

**Comment by IamPete1** [comment]
@rishabsingh3003 I think we can do this now?

**Comment by rishabsingh3003** [comment]
Yes, it is! Via OBSTACLE_DISTANCE_3D mavlink message. Fences are still 2D though .. or rather Horizontal + Vertical

**Comment by rmackay9** [comment]
TBH, I don't think that the obstacle-distance-3d is really for the purpose specified by this issue.

**Comment by peterbarker** [comment]
On Sun, 15 Aug 2021, Randy Mackay wrote:

> TBH, I don't think that the obstacle-distance-3d is really for the purpose specified by this issue.

Agreed.  It's more like the opposite of the mavlink SAFETY_ALLOWED_AREA 
concept.  Which is a volume, not an area....

**Comment by peterbarker** [comment]
.... but @rmackay9 's suggestion of alt limits for separate polygon fences is probably a better way forward.  if rather computationally expensive.

**Comment by manavgandhi17** [comment]
Pinging back on this to see if there has been any developmental work done on fence-based 3D obstacle avoidance. Would be very helpful for sectioning out trees instead of an infinite vertical cylinder.

We would probably also need to integrate altitude type into this distinction (MSL, AGL, Terrain) - as well has somehow configuring the same altitude value/type for all vertices of a singular polygon.
